### PR TITLE
MappingEffect, Change, dialog

### PR DESCRIPTION
**Pull request by BurntPizza**
Not thoroughly tested, looking for some review.
Vast majority of the code is UI, actual work is in MappingChange + PiecewiseLinearMapping.
Big chunk of it is making it pretty...
Not sure if some of the helper classes should be split out.

Also doesn't currently do anything more than piece-wise linear maps like splines, etc. although those can be fairly easily added.

Checkboxes are left-click for normal toggle, right-click for solo+all-others-toggle, middle-click to enable all. Also right-click on points to remove.

(Pardon the low-framerate gif)

Imgur

### CODE REVIEW

File: src/heroesgrave/spade/core/effects/MappingEffect.java [hunk 1]
```diff
@@ -0,0 +1,421 @@
...
+			return start.y + (end.y - start.y) * ((x - start.x) / (end.x - start.x));
+		}
+		
+		static PiecewiseLinearMapping identity() {
+			return new PiecewiseLinearMapping();
+		}
+	}
+	
+	// something something tuples
+	private static class Pair<A, B> {
```
Comment (HeroesGrave) [inline src/heroesgrave/spade/core/effects/MappingEffect.java]: There's already a Pair class in [`heroesgrave.utils.misc`](https://github.com/Spade-Editor/Spade/blob/master/src/heroesgrave/utils/misc/Pair.java)

**Comment by HeroesGrave** [comment]
Looks good to me. I'll give it a bit of a test this weekend and see if I can break it ;)

Should be fine to merge as is, unless you want to make some more adjustments.

File: src/heroesgrave/spade/core/effects/MappingEffect.java (hunk 1 above)
Comment (BurntPizza) [inline src/heroesgrave/spade/core/effects/MappingEffect.java]: Done.

**Comment by BurntPizza** [comment]
What do you think of https://github.com/BurntPizza/Core/commit/657d0dbece497f3b45c435a5fb0d23eb26b5d17c?
Also found a little bug (not fixed): dragging a point on top of a fixpoint makes the line go black instead of staying just gray.

**Comment by BurntPizza** [comment]
Also, is this the correct way to do this? https://github.com/BurntPizza/Core/blob/master/src/heroesgrave/spade/core/effects/MappingEffect.java#L64
The mapping should be specific to the layer, but isn't this applying it to all layers in the doc?

**Comment by BurntPizza** [comment]
Spectral view now updates with the graph.
Imgur

**Comment by BurntPizza** [comment]
WRT https://github.com/BurntPizza/Core/commit/c0448ac9f98825b47df06623d17811d44acee7c9, what would be the way to go about updating the MappingEffect dialog when the selection changes?

**Comment by HeroesGrave** [comment]
> Also, is this the correct way to do this? https://github.com/BurntPizza/Core/blob/master/src/heroesgrave/spade/core/effects/MappingEffect.java#L64 Also line 111.
The mapping should be specific to the layer, but isn't this applying it to all layers in the doc?

Yep. That's the correct way to do it, even if it seems a little bit strange. I suppose I could make it a bit more intuitive by wrapping it in a method in the Layer class.

> WRT BurntPizza@c0448ac, what would be the way to go about updating the MappingEffect dialog when the selection changes?

There currently isn't any mechanism to detect when the image (or selection) has been changed. IMO, the sensible solution would be to just disallow changes while previewing or configuring an effect.

**Comment by BurntPizza** [comment]
Got it back to merge status, you can probably go ahead.

**Comment by HeroesGrave** [comment]
Hmm. I thought a simple rebase would've worked fine.

Thanks.

**Pull request merged**